### DISCUSSION TITLE
What is the purpose of `extraLuaPackages`?

### DISCUSSION THREAD

**Discussion by nicolas-goudry** [nix-community/nixvim#1700] [Q&A]
Just wondering what is the purpose of the `extraLuaPackages` option?

I tried to use it to add Lua modules to my Nixvim config, but the modules are not available in Neovim…

For example, I once tried to add the [inspect](https://github.com/kikito/inspect.lua) module in order to be able to pretty print tables in Neovim, but it didn’t work.

```nix
{
  extraLuaPackages = ps: [ ps.inspect ];
}
```

To be clear, Neovim works fine, but if I try to run `:lua print(require("inspect")({ test = "test" }))` it just fails with the following error message:

```
E5108: Error executing lua [string ":lua"]:1: module 'inspect' not found:
```

**Comment by MattSturgeon** [comment]
Good question. It's defined here, but I can't find the actual implementation...

https://github.com/nix-community/nixvim/blob/7a2a25af02be25987aa43cd681312f4b5ba12317/modules/output.nix#L93

Based on uses, I'd say it's a wrapper on `extraPackages` with access to the `pkgs.luaPackages` scope used to build neovim.

Therefore, adding luaPackages is unlikely to add the corresponding vim plugins to your runtimepath.

Rather than adding `pkgs.luaPackages.*` to your `extraPackages` you should be adding `pkgs.vimPlugins.*`.

Either use `extraPlugins` to define plugin packages and config together, or `extraPackages` to just install the plugin package.

**Reply by nicolas-goudry** [reply]
Thanks for your answer :slightly_smiling_face: 

I did find the option definition, but like you I couldn’t find the implementation… Nevertheless, I think your explanation does make sense. But then it seems kind of impossible (currently at least) to just add a random Lua module to the runtimepath.

Should I open an issue in this regard?

**Reply by MattSturgeon** [reply]
The first couple [FAQs](https://nix-community.github.io/nixvim/user-guide/faq.html) are on this topic. Do they help clarify anything?

Also, do you have any feedback on (specific parts of) the docs? We're always looking to improve them. Specific actionable feedback is the most helpful, but we'll take anything 😅

You're welcome to open separate issues/discussions, chat on matrix or continue in this thread.

**Reply by nicolas-goudry** [reply]
Since `inspect` is not a Vim plugin per-se (and neither is packaged as a Vim plugin), sadly the FAQ doesn’t help in this regard.

As for feedback, I’m pretty happy with the way things are done in Nixvim and the documentation is pretty great! However, we could always improve, so here are a few nitpicks I can think of right away:

- deprecated options are not marked as such in documentation
- options discovery could be better (search? like on search.nixos.org) since the options sub-attrsets are in a different page

Another thing, which seems like a good idea but may very well not be one:

I often need to define keymaps for some plugins which do not provide a `keymaps` option so it would be great to be able to add a `keymaps` option to all plugins that do not provide one OOTB. This `keymaps` would then be merged with the global `keymaps` option.

For context, here is my config: https://github.com/nicolas-goudry/nixvim-config

As you can see, I split my config in « topics » and made an [auto-loader for plugins](https://github.com/nicolas-goudry/nixvim-config/blob/main/plugins/default.nix). Then, in the plugins I define `opts` which are the plugin options and I can use `rootOpts` to define root options, like `keymaps`. [Example](https://github.com/nicolas-goudry/nixvim-config/blob/main/plugins/smart-splits.nix)

I have the feeling that this could be a pretty common use case, but I also fear that it would complexify Nixvim configuration understanding/discovery.
